ldevnet/ldevnet.go

```diff
@@ -0,0 +1,326 @@
+package ldevnet
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"io/ioutil"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-sectorbuilder"
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/api/apistruct"
+	"github.com/filecoin-project/lotus/api/client"
+	"github.com/filecoin-project/lotus/api/test"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/actors"
+	"github.com/filecoin-project/lotus/chain/gen"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/genesis"
+	"github.com/filecoin-project/lotus/lib/jsonrpc"
+	"github.com/filecoin-project/lotus/miner"
+	"github.com/filecoin-project/lotus/node"
+	"github.com/filecoin-project/lotus/node/modules"
+	modtest "github.com/filecoin-project/lotus/node/modules/testing"
+	"github.com/filecoin-project/lotus/node/repo"
+	"github.com/filecoin-project/lotus/storage/sbmock"
+	"github.com/ipfs/go-datastore"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	peer "github.com/libp2p/go-libp2p-peer"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func init() {
+	build.SectorSizes = []uint64{1024}
+	build.MinimumMinerPower = 1024
+}
+
+type LocalDevnet struct {
+	Client *apistruct.FullNodeStruct
+	Miner  address.Address
+
+	numMiners int
+	cancel    context.CancelFunc
+	closer    func()
+	done      chan struct{}
+}
+
+func (ld *LocalDevnet) Close() {
+	ld.cancel()
+
+	for i := 0; i < ld.numMiners; i++ {
+		<-ld.done
+	}
+	close(ld.done)
+	ld.closer()
+}
+
+func New(t *testing.T, numMiners int) (*LocalDevnet, error) {
```

Comment:
This package `ldevnet` is for building the temp FIL devnet. 
Mostly based on _TestDealFlow_ test from Lotus, mixed with other test to enable API... so to make a FIL network with a client Lotus node, and let our code communicate via API (not internal things). So, to simulate what we're really going to do.

I stretched a bit for supporting more than 1 miner compared to original impl... seems to work but not 100% sure, have to play more with this to be sure.